text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+
+    Error: could not serialize access due to concurrent update
+
+It's expected that transaction may be repeated and be successful when
+repeated. Common case when this happens is using ``INSERT ... UNLESS
+CONFLICT ..`` statement in EdgeDB. See `Repeating-EdgeQL<#repeating-edgeql>`_
+section for an explanation of why Postgres and EdgeDB can't repeat transactions
+automatically.
+
+EdgeDB has only ``REPEATABLE READ`` and ``SERIALIZABLE`` isolation levels.
+That means errors like the above may happen more often than with the
+default ``READ COMMITTED`` isolation level in Postgres.
+
+As we implement retrying connection on "concurrent update" we also want
+to handle:
+
+1. Transaction deadlocks
+2. Connection errors anywhere between ``BEGIN TRANSACTION`` and before
+   ``COMMIT`` (we can't reliably know whether commit itself worked if we sent
+   last ``COMMIT`` and got no response)
+
+`More discussion on retrying <https://github.com/edgedb/edgedb/discussions/1738>`_
+
+
+Specification
+=============
+
+General idea of the feature:
+
+1. Introduce a block of user's code that may be repeated
+2. Pass a transaction object to the block
+3. Repeat the block within a new transaction until succeeds or attempt
+   number is exhausted
+
+The block is introduced by ``connection_pool.retry``, the exact code
+is different for different languages because of language limitations
+and convetions.
+
+Normal transactions that aren't retried are executed with
+``try_transaction`` method.
+
+The ``retry`` function configured by the number of attempts and delay function.
+User can set either number of retry attempts either globally (for all errors)
+or for specific error condition (say only deadlocks).
+
+Current attempt number N is global. Which means if the last error is a
+deadlock and N is greater than the number of attempts on a deadlock we
+stop retrying and return error (even if previous error was a network
+error).
+
+Delay function by default is ``2^N * 100`` plus random number in range
+``0..100`` microseconds. Where first retry (second attempt) has ``N=1``.
+Technically:
+
+* In JavaScript: ``n => (2**n) * 100 + Math.random()*100``
+* In Python: ``lambda n: (2**n) * 0.1 + randrange(100)*0.1`` (seconds)
+* In Rust: ``|n| Duration::from_millis(2u64.pow(n)*100 + thread_rng().gen_range(0,100)``
+
+Delay is randomized so that if there was a coordinated failure (i.e. server
+restart which triggers all current transactions to retry) transactions don't
+overwhelm a database by reconnecting simultaneously. If delay function is
+adjusted it's recommeded to keep some randomization anyway.
+
+
+TypeScript API
+--------------
+
+**Introduce** two methods on a connection pool and one on connection:
+
+.. code-block:: typescript
+
+    type TransactionBlock = (Transaction) => Promise<T>;
+
+    interface Pool {
+        retry<T>(block: TransactionBlock, options: RetryOptions): T;
+        raw_transaction<T>(action: TransactionBlock, options?: TransactionOptions): T;
```

Comment:
```suggestion
        try_transaction<T>(action: TransactionBlock, options?: TransactionOptions): T;
```

?

Reply:
Yes, I like `try` more than `raw`, as the former term plays nicely with `retry`. Also, all JS APIs should use camelCase, so it's

```suggestion
        tryTransaction<T>(action: TransactionBlock, options?: TransactionOptions): T;
```